src/state_manager/mod.rs

```diff
@@ -1230,53 +1252,77 @@ where
-        mut ts: Arc<Tipset>,
-        height: i64,
-    ) -> Result<(), anyhow::Error> {
-        if height > ts.epoch() {
-            anyhow::bail!(
-                "height {height} cannot be greater than tipset epoch {}",
-                ts.epoch()
-            );
-        }
-        let mut ts_chain = Vec::<Arc<Tipset>>::new();
-        while ts.epoch() != height {
-            let next = self.cs.tipset_from_keys(ts.parents())?;
-            ts_chain.push(std::mem::replace(&mut ts, next));
-        }
-        ts_chain.push(ts);
-
-        let mut last_state = *ts_chain.last().unwrap().parent_state();
-        let mut last_receipt = *ts_chain.last().unwrap().blocks()[0].message_receipts();
-        for ts in ts_chain.iter().rev() {
-            if ts.parent_state() != &last_state {
-                anyhow::bail!(
-                    "Tipset chain has state mismatch at height: {}, {} != {}, \
-                        receipts mismatched: {}",
-                    ts.epoch(),
-                    ts.parent_state(),
-                    last_state,
-                    ts.blocks()[0].message_receipts() != &last_receipt
-                );
-            }
-            if ts.blocks()[0].message_receipts() != &last_receipt {
-                anyhow::bail!(
-                    "Tipset message receipts has a mismatch at height: {}",
-                    ts.epoch(),
-                );
-            }
-            info!(
-                "Computing state (height: {}, ts={:?})",
-                ts.epoch(),
-                ts.cids()
-            );
-            let (st, msg_root) = self.tipset_state(ts).await?;
-            last_state = st;
-            last_receipt = msg_root;
-        }
-        Ok(())
+    /// Validates all tipsets at epoch `start..=end` behind the heaviest tipset.
+    ///
+    /// This spawns [`rayon::current_num_threads`] threads to do the compute-heavy work
+    /// of tipset validation.
+    ///
+    /// # What is validation?
+    /// Every state transition returns a new _state root_, which is typically retained in e.g snapshots.
+    /// For "full" snapshots, all state roots are retained.
+    /// For standard snapshots, the last 2000 or so state roots are retained.
+    ///
+    /// _receipts_ meanwhile, are typically ephemeral, but each tipset knows the _receipt root_
+    /// (hash) of the previous tipset.
+    ///
+    /// This function takes advantage of that fact to validate tipsets:
+    /// - `tipset[N]` claims that `receipt_root[N-1]` should be `0xDEADBEEF`
+    /// - find `tipset[N-1]`, and perform its state transition to get the actual `receipt_root`
+    /// - assert that they match
+    ///
+    /// See [`Self::compute_tipset_state_blocking`] for an explanation of state transitions.
+    ///
+    /// # Known issues
+    /// This function is blocking, but we do observe threads waiting and synchronizing.
+    /// This is suspected to be due to the VM's use of the chain store, or maybe JIT-ing.
```

Comment:
What kind of JIT?

Reply:
Not sure, taken from this comment. I'll reword
https://chainsafeio.slack.com/archives/C0144LHN732/p1687512425630809?thread_ts=1687471388.702309&cid=C0144LHN732